uaclient/cli/commands.py

```diff
@@ -0,0 +1,104 @@
+import argparse
+from typing import Callable, List, Optional, Union
+
+from uaclient.cli.constants import NAME, USAGE_TMPL
+
+
+class ProArgument:
+    def __init__(
+        self,
+        long_name: str,
+        help: str,
+        short_name: Optional[str] = None,
+        **kwargs
+    ):
+        self.names = (
+            (long_name,) if short_name is None else (short_name, long_name)
+        )
+        self.help = help
+        self.additional_args = kwargs
+
+    def register(
+        self, target: Union[argparse.ArgumentParser, argparse._ArgumentGroup]
+    ):
+        target.add_argument(
+            *self.names, help=self.help, **self.additional_args
+        )
+
+
+class ProArgumentMutuallyExclusiveGroup:
+    def __init__(
+        self,
+        required: bool = False,
+        arguments: List[ProArgument] = [],
+    ):
+        self.required = required
+        self.arguments = arguments
+
+
+class ProArgumentGroup:
+    def __init__(
+        self,
+        title: Optional[str] = None,
+        description: Optional[str] = None,
+        arguments: List[ProArgument] = [],
+        mutually_exclusive_groups: List[
+            ProArgumentMutuallyExclusiveGroup
+        ] = [],
+    ):
+        self.title = title
+        self.description = description
+        self.arguments = arguments
+        self.mutually_exclusive_groups = mutually_exclusive_groups
+
+    def register(self, parser: argparse.ArgumentParser):
+        target = (
+            parser
+        )  # type: Union[argparse.ArgumentParser, argparse._ArgumentGroup]
+        if self.title:
+            target = parser.add_argument_group(self.title, self.description)
+
+        for argument in self.arguments:
+            argument.register(target)
+
+        for group in self.mutually_exclusive_groups:
+            new_group = target.add_mutually_exclusive_group(
+                required=group.required
+            )
+            for argument in group.arguments:
+                argument.register(new_group)
+
+
+class ProCommand:
+    def __init__(
+        self,
+        name: str,
+        help: str,
+        description: str,
+        usage: Optional[str] = None,
+        action: Callable = lambda *args, **kwargs: None,
+        preserve_description: bool = False,
+        argument_groups: List[ProArgumentGroup] = [],
```

Comment:
nit (that applies in a couple places in this file): default values of mutable lists can lead to confusing behavior. If one call modifies the default value of `argument_groups`, then all calls will get that modified default value. Maybe make it `Iterable[ProArgumentGroup]` and set an empty tuple as the default value?

Reply:
aha good catch I forgot about that